### DISCUSSION TITLE
Contributing as a University Student

### DISCUSSION THREAD

**Discussion by mcnojo** [networkx/networkx#7234] [Q&A]
Hello NX Folks!

I'm Noah McGuire, a computer science (and cognitive science) student at the University of Michigan.  I'm currently taking a software engineering class and our final assignment this semester involves submitting a non trivial modification or extension of an open source project and documenting our process to the standards of the assignment.

Though the due date is months away, I think this is the perfect repo for me to contribute to because I will be making use of it for a separate research project involving knowledge graphs and the analysis of their embedding models, and so will be able to really sink deeply into the codebase.  

Naturally, I'm willing to try and address the present issues, but I wanted to ask if would be allowable for me to experiment with (and potentially add to NX!) a set of functions to perform analysis of topological artifacts and dynamics of higher-order multiplex networks.

I see that there was a previous project by Vadim Abzalov for making a topological sorting notebook, so I figured it would be valuable to ask if this is a feasible pursuit!

If this doesn't seem feasible or you are aware of other projects where this pursuit would be better places, please let me know!

Thanks!

**Comment by dschult** [comment]
These "higher-order multiplex networks" used to be called hypergraphs. And NetworkX doesn't have algorithms that deal with them. I think we would need new data structures (unless the problems can be reduced to 2-nodes-per-edge in some clever way). It'd be better for you to focus on learning about various analysis techniques for those problems. If there is a feature that fits within NetworkX great! But I wouldn't recommend counting completing that and getting it reviewed and edited and changed, etc. for a one semester term project. Knowledge graphs and embedding models could be a better fit -- but a lot depends on you finding a good problem to sink your teeth into. In terms of a good project for your software course, you want it to be self-contained and well-defined. Such algorithms do exist and you may come across one in your analysis course.

On the more software side (if the analysis projects end up too difficult), if you want to dive into python (and one of its new features) take a look at [PR 7166](https://github.com/networkx/networkx/pull/7166) which switches some code from using many `if/elif/else` clauses into using the new `match` feature.  That PR is about the shortest_path functions. But there are other places in NetworkX with long if/elif/else clauses that could be changed. 

Take a look around -- maybe something completely different will strike your fancy. :}
